src/libgit2/transports/emscriptenhttp.cc

```diff
@@ -0,0 +1,271 @@
+#ifdef __EMSCRIPTEN__
+
+#include "emscripten.h"
+#include "emscripten/fetch.h"
+
+extern "C" {
+#include "common.h"
+#include "git2/transport.h"
+#include "smart.h"
+}
+
+#include <map>
+
+#include "deps/picosha2/picosha2.h"
+
+static const char *upload_pack_ls_service_url = "/info/refs?service=git-upload-pack";
+static const char *upload_pack_service_url = "/git-upload-pack";
+static const char *receive_pack_ls_service_url = "/info/refs?service=git-receive-pack";
+static const char *receive_pack_service_url = "/git-receive-pack";
+
+namespace {
+struct StreamInternal {
+  std::string url;
+  std::vector<char> writeBuffer;
+  std::vector<const char *> headers;  // [key1, value1, key2, value2, ...]
+  emscripten_fetch_attr_t attr;
+  emscripten_fetch_t *fetch{nullptr};
+  size_t totalBytesRead{0};
+};
+
+uint64_t connectionCount{0};
+std::map<uint64_t, StreamInternal> connectionMap;
+
+uint64_t xhrConnect(
+  const std::string url, const char *method, const std::vector<const char *> headers) {
+  auto connectionNumber = connectionCount++;
+  auto &connection = connectionMap[connectionNumber];
+  connection.url = url;
+  connection.headers = headers;
+  emscripten_fetch_attr_init(&connection.attr);
+  strcpy(connection.attr.requestMethod, method);
+  // NOTE(pawel) EMSCRIPTEN_FETCH_REPLACE is needed for synchronous to work...
+  // https://github.com/emscripten-core/emscripten/issues/8183
+  connection.attr.attributes =
+    EMSCRIPTEN_FETCH_LOAD_TO_MEMORY | EMSCRIPTEN_FETCH_SYNCHRONOUS | EMSCRIPTEN_FETCH_REPLACE;
+  if (std::string(method) == "GET") {
+    auto headersToSend = connection.headers;
+    headersToSend.push_back(0);
+    connection.attr.requestHeaders = headersToSend.data();
+    connection.fetch = emscripten_fetch(&connection.attr, url.c_str());
+  }
+  return connectionNumber;
+}
+
+// This call buffers the writes. The buffer is sent on the wire when xhrRead() is called.
+void xhrWrite(uint64_t connectionNumber, const char *buffer, size_t size) {
+  if (connectionMap.count(connectionNumber) != 1) {
+    printf("Attempting to write to connection %l but it is not connected", connectionNumber);
+    return;
+  }
+  auto &connection = connectionMap[connectionNumber];
+  connection.writeBuffer.insert(end(connection.writeBuffer), buffer, buffer + size);
+}
+
+// Sends pending writes and returns response async. The result is buffered so this can be invoked
+// until the full length of the buffer is read.
+void xhrRead(uint64_t connectionNumber, char *buffer, size_t bufferSize, size_t *bytesRead) {
+  if (connectionMap.count(connectionNumber) != 1) {
+    printf("Attempting to read from connection %l but it is not connected\n", connectionNumber);
+    *bytesRead = 0;
+    return;
```

Comment:
Should we be returning `-1` here?

Reply:
It would probably be a good thing to do. And I'll pass through this error throw the relative `emscriptenhttp_stream_read` and `emscriptenhttp_stream_write_single`